### PR TITLE
add varlen attention for qwen 3

### PR DESCRIPTION
**Pull request by liangel-02**
As title 

**Testing**

Screenshot 2025-11-24 at 4 30 53 PM

performance and loss on par

### CODE REVIEW

File: torchtitan/models/qwen3/model/model.py [hunk 1]
```diff
@@ -170,6 +173,8 @@ def __init__(self, model_args: Qwen3ModelArgs):
         match self.attn_type:
             case "flex":
                 self.inner_attention = FlexAttentionWrapper()
+            case "varlen":
+                self.inner_attention = VarlenAttentionWrapper()
             case _:
```
Comment (tianyu-l) [inline torchtitan/models/qwen3/model/model.py]:
```suggestion
            case "sdpa":
```
and raise ValueError as fallback

File: torchtitan/models/qwen3/model/model.py [hunk 2]
```diff
@@ -231,6 +236,15 @@ def forward(
...
+            case "varlen":
+                assert isinstance(attention_masks, VarlenMetadata), attention_masks
+                output = self.inner_attention(
+                    xq,
+                    xk,
+                    xv,
+                    self.head_dim,
+                    attention_masks,
+                )
             case _:
```
Comment (tianyu-l) [inline torchtitan/models/qwen3/model/model.py]: same

**Comment by tianyu-l** [comment]
also need to add sac support for Qwen3 varlen

**Comment by liangel-02** [comment]
@wwwjn yes, i think compiling both is causing the slight difference in loss. this is the loss when both are not compiled which ends up being a lot closer 

Screenshot 2025-11-25 at 12 31 12 PM

cc @drisspg

**Review by wwwjn — APPROVED** [review]
LGTM!

**Review by tianyu-l — APPROVED** [review]
